src/QueueBuffer.mo

```diff
@@ -0,0 +1,242 @@
+  };
+
+  // Buffer
+  //
+  // A linear buffer where we can add at end and delete from the beginning.
+  //
+  // This data structure consists of a pair of Vectors called `old` and `new`.
+  // We always add to `new`.  While `old` is empty we delete from `new` but only
+  // until the waste in `new` exceeds sqrt(n). When `new` has >sqrt(n) waste
+  // then we rename `new` to `old` and create a fresh empty `new`. Now deletions
+  // happen from old, until old is empty. Then `old` is discarded and deletions
+  // happen from `new` again until the waste in `new` exceeds sqrt(n). Then the
+  // shift starts all over again. Etc.
+  //
+  // Only the waste in `new` is limited to sqrt(n). The waste in `old` is not limited.
+  // Hence, the largest waste occurs if we do n additions first, then n deletions.
+  class Buffer<X>() {
+
+    var old : ?Vector<X> = null;
+    var new : Vector<X> = Vector<X>();
+    var n = 0; // offset of old
+    var m = 0; // offset of new
+
+    public func add(x : X) : Nat {
+      new.add(x) + m;
+    };
+
+    public func getOpt(i : Nat) : ?X {
+      if (i >= m) {
+        new.getOpt(i - m : Nat);
+      } else if (i >= n) {
+        let ?vec = old else Prim.trap("old is null in Buffer");
+        vec.getOpt(i - n : Nat);
+      } else null;
+    };
+
+    func rotateIfNeeded() {
+      let size = new.size();
+      let s = Nat32(size);
+      let d = Nat32(new.deleted());
+      let bits = 32 - leadingZeros(s);
+      let limit = s >> (bits >> 1);
+      if (d > limit) {
+        old := ?new;
+        n := m;
+        new := Vector<X>();
+        m := n + size;
+      };
+    };
+
+    public func delete(n_ : Nat) {
+      var ctr = n_;
+      let end = deleted() + ctr;
+      if (end > size()) Prim.trap("index out of bounds in Buffer.delete");
+      switch (old) {
+        case (?vec) {
+          if (vec.size() > ctr) {
+            vec.delete(ctr);
+            return;
+          } else {
+            ctr := ctr - vec.size();
+            old := null;
+          };
+        };
+        case (null) {};
+      };
+      new.delete(ctr);
+      rotateIfNeeded();
+    };
+
+    public func deleted() : Nat = switch (old) {
+      case (?vec) { n + vec.deleted() };
+      case (null) { m + new.deleted() };
+    };
+
+    public func size() : Nat = m + new.size();
+    public func len() : Nat = size() - deleted();
+  };
+
+  /// A queue with fast random access, which preserves history of popped values
+  public class QueueBuffer<X>() {
+    var buf : Buffer<X> = Buffer<X>();
+    var head : Nat = 0;
+
+    /// get id of oldest item in queue history
+    public func histId() : Nat = buf.deleted();
+    /// get id of oldest item in queue
+    public func headId() : Nat = head;
+    /// get next id which will be issued
+    public func nextId() : Nat = buf.size();
+
+    /// amount of items in the queue
+    public func size() : Nat = buf.size() - head;
```

Comment:
Can we call this queueSize()?

It's been hard to remember what size() refers to.